components/architecture.md

```diff
@@ -0,0 +1,147 @@
+
+### Components
+
+A component is an isolated set of functionality that is likely useful in its
+own right.
+
+A component:
+- MUST provide a GRPC endpoint with the following services on a node-local socket (such as a unix socket or named pipe):
+	- MUST provide [GRPC healthcheck service](https://github.com/grpc/grpc/blob/master/doc/health-checking.md)
+	- MUST provide [Version service](api/services/component/version/v1)
+	- SHOULD provide [Metrics service](api/services/component/metrics/v1)
+- MAY expose a separate API for external consumption
+- SHOULD use GRPC to publish component-level services
+- SHOULD be easily deployed as a container (the primary packaging format for
+Moby Core, see [Packaging](#Packaging)).
+	- SHOULD log to stdout and/or stderr
+	- SHOULD NOT depend on shared filesystem access with other components
+		- If shared filesystem access is required, this SHOULD be well documented
+
+### Moby Core Services
+
+Moby Core defines a base set of services to be used with components along with
+default implementations of those services.
+A "service" in this case refers to a [GRPC service](https://grpc.io/docs/guides/concepts.html#service-definition).
+A component may consist of one or more services.
+
+All Moby Core services MUST use GRPC for cross-component communication.
+Proto files MUST be separate from the implementation, however generated
+code MUST stay with the proto.
+
+While these services are split up here, multiple services MAY be part of
+a single component.
+
+#### Public API's
+
+- moby-http-api-v1 - Provides the v1.* HTTP API
+	- Dependencies:
+		- moby-container-store
+		- moby-container-rootfs
+		- moby-plugin-store
+		- moby-network-store
+		- moby-volume-store
+		- moby-image-store
+
+#### Events
+
+- moby-events - Pub/sub for component level events.
+
+#### Container management
+
+- moby-container-store - Provides basic CRUD operations for container
+specs/metadata. May be unnecessary or at least reduced scope after
+[containerd#1378](https://github.com/containerd/containerd/pull/1378)
+- moby-container-runtime - start/stop/kill/etc operations on containers
+	- Dependencies:
+		- moby-network-runtime
+		- moby-volume-runtime
+- moby-container-rootfs - Provides access to container filesystems.
+- moby-container-logs - Collects container stdio and forwards to a centralized logging service
+
+#### Plugins
+
+- moby-plugin-store - Provides basic CRUD operations for plugins
+
+#### Networking
+
+- moby-network-store - Provides basic CRUD operations for networks
+	- Dependencies:
+		- moby-plugin-store
+- moby-network-runtime - Sets up networking for containers
+	- Dependencies:
+		- moby-network-store
+
+#### Persistent storage
+
+- moby-volume-store - Provides basic CRUD operations for volumes
+	- Dependencies:
+		- moby-plugin-store
+- moby-volume-runtime - Sets up volumes for a container
+	- Dependencies:
+		- moby-volume-store
+
+#### Image Management
+
+- moby-layer-store - Storage for image layers
```

Comment:
Is there a motivation for having this be a separate component? We want this part of the code to be deprecated by containerd and only have containerd implement the image store interface for doing image management. The implementations themselves will already have sub-components but does that make it a "moby component"?

Reply:
So this wouldn't be the literal layer store that's currently in moby/moby.

This may not be a component by itself, but rather part of one (a component may comprise of more than 1 service). But it's something we need to implement.
What will actually be in here I'm not sure yet, but we need to support the existing drivers (even if re-written to be snapshotters) at least.